### PR TITLE
Introduce new package ghc-lib-parser

### PR DESCRIPTION
**Pull request by shayne-fletcher-da**
In this PR, `ghc-lib` has been refactored into two packages : `ghc-lib-parser` and `ghc-lib` with the latter depending on the former.

### CODE REVIEW

File: CI.hs [hunk 1]
```diff
@@ -13,34 +13,34 @@ import System.Time.Extra
...
-
-    -- Generate an sdist and extract it to ensure it works.
+    cmd "stack exec -- ghc-lib-gen ghc --ghc-lib-parser"
     stackYaml <- readFile' "stack.yaml"
     writeFile "stack.yaml" $ stackYaml ++ unlines ["- ghc"]
-    cmd "stack sdist ghc --tar-dir=."
-    [tarball] <- filter (isExtensionOf ".tar.gz") <$> getDirectoryContents "."
-    cmd $ "tar -xf " ++ tarball
+    tarball <- sDistCreateExtract
+    renameDirectory (dropExtension $ dropExtension tarball) "ghc-lib-parser"
```
Comment (neil-da) [inline CI.hs]: `dropExtensions` ?

File: CI.hs [hunk 2]
```diff
@@ -13,34 +13,34 @@ import System.Time.Extra
...
+    removeFile tarball
+    removeFile "ghc/ghc-lib-parser.cabal"
+
+    -- Make and extract an sdist of ghc-lib.
+    cmd "cd ghc && git checkout . && cd .."
```
Comment (neil-da) [inline CI.hs]:
```suggestion
    cmd "cd ghc && git checkout ."
```

These run in separate shells - changing directory back after is pointless

File: ghc-lib-gen/src/Main.hs [hunk 3]
```diff
@@ -251,7 +457,85 @@ generateCabal = do
...
+    bin <- (:[]) <$> readCabalFile cabalFileBinary
+    let askField from x = nubSort $ concatMap (`askCabalField` x) from
+    let askFiles from x = nubSort $ concatMap (`askCabalFiles` x) from
+
+    writeFile "ghc-lib-parser.cabal" $ unlines $ map trimEnd $
+        -- header
+        ["cabal-version: >=1.22"
+        ,"build-type: Simple"
+        ,"name: ghc-lib-parser"
+        ,"version: 0.1.0"
```
Comment (neil-da) [inline ghc-lib-gen/src/Main.hs]: I assume we then manually change the version?

Comment (shayne-fletcher-da) [reply ghc-lib-gen/src/Main.hs]: Yes.

**Pull request merged**